doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+
+Replica IDs are global between all clusters, so introspection tables can be
+uniquely named by log variant and replica ID:
+
+```text
+<mz-log-variant>_<replica-id>
+```
+
+... where `<mz-log-variant>` is the value of `BuiltinLog::name` for the
+respective log variant. For example, for log variant `Timely(Operates)` and
+replica `5` the introspection table name is `mz_dataflow_operators_5`.
+
+When a replica is dropped, the coordinator removes all associated introspection
+tables.
+
+Query planning translates references to introspection tables to reads from
+persist sources using `ExternalSourceConnector::Persist` and the respective
+persist shard ID for that log variant-replica combination.
+
+Introspection tables being backed by unmaterialized sources means that they have
+worse ergonomics that the existing arrangement-backed introspection tables. To
+query them, users must either specify an explicit `AS OF` or create an index. We
+consider this loss in ergonomics acceptable.
+
+#### Introspection Views
+
+For convenient access to introspection data from all replicas, the coordinator
+maintains a built-in introspection view for each log variant. An introspection
+view contains the union of all replica-specific introspection tables for the
+respective log variant, tagged with cluster name and replica name:
+
+```sql
+CREATE VIEW <mz-log-variant>_persist AS
+  (
+    SELECT 'c1' AS cluster, 'r1' AS replica, *
+    FROM <mz-log-variant>_1
+  ) UNION (
+    SELECT 'c1' AS cluster, 'r2' AS replica, *
+    FROM <mz-log-variant>_2
+  ) UNION (
+    SELECT 'c2' AS cluster, 'r1' AS replica, *
+    FROM <mz-log-variant>_3
+  ) ...
+```
+
+Note that the `_persist` (or some other) suffix to the view name is required
+because the old arrangement-based introspection tables already occupy the plain
+`<mz-log-variant>` names. Once we remove the arrangement-based introspection
+tables, we can also drop the suffix.
+
+When a new replica is created, the coordinator adds its introspection tables to
+the respective introspection views.
+
+When a replica is dropped, the coordinator removes its introspection tables from
+the respective introspection views.
+
+### Feature Flag
+
+As mentioned previously, we don't want to enable logging to persist by default
+as long as persist does not support compaction, as otherwise we would risk
+unbounded growth of the introspection collections. Therefore, logging to
+persist is disabled by default.
+
+We add a feature flag `--replicated-compute-introspection` to both
+`materialized` and `computed`, to opt into logging introspection data to
+persist. `materialized` passes the flag (if set) on to `computed`.
+
+If the flag is not set, the coordinator does not create the introspection tables
+and views described above. It may still generate shard IDs and pass them to the
+compute controller, as this is invisible to users.
+
+If the flag is not set, replicas don't sink introspection data to persist.
```

Comment:
I'm not sure if this is the right way to opt-in to features in Materialize. Do we already have precedence for doing something like this?

Reply:
Ah, there was `--persistent-user-tables`, which was removed in https://github.com/MaterializeInc/materialize/pull/11834.